femtools/Field_derivatives.F90

```diff
@@ -42,16 +42,280 @@ module field_derivatives
+        end if
+
+        if (present(bc_value)) then
+          call dg_ele_grad_scalar_interface(ele, face, face_2, ni, &
+               & loc_grad, X, field, bc_value, bc_type)
+        else
+          call dg_ele_grad_scalar_interface(ele, face, face_2, ni, &
+               & loc_grad, X, field)
+        end if
+      end do
+
+      ! multiply by inverse of mass matrix
+      inv_mass = inverse(shape_shape(shape, shape, detwei))
+      do i = 1, mesh_dim(field)
+        loc_grad(i,:) = matmul(inv_mass, loc_grad(i,:))
+      end do
+
+    end function dg_ele_grad_scalar
+
+    subroutine dg_ele_grad_scalar_interface(ele, face, face_2, &
+         ni, loc_grad, X, field, bc_value, bc_type)
+
+      !!< Construct the DG element boundary integrals on the ni-th face of
+      !!< element ele.
+      integer, intent(in) :: ele, face, face_2, ni
+      type(scalar_field), intent(in) :: field
+      type(vector_field), intent(in) :: X
+      type(scalar_field), intent(in), optional :: bc_value
+      integer, dimension(:), intent(in), optional :: bc_type
+      real, dimension(mesh_dim(field), ele_loc(field, ele)), intent(inout) :: loc_grad
+
+      ! Face objects and numberings.
+      type(element_type), pointer :: shape
+      real, dimension(mesh_dim(field), face_ngi(field, face)) :: normal
+      real, dimension(face_ngi(field, face)) :: detwei, in_q, in_q_2, in_bc_q
+      real, dimension(mesh_dim(field), face_ngi(field, face)) :: vector
+      real, dimension(mesh_dim(field), face_loc(field, face)) :: face_rhs
+      real, dimension(ele_loc(field, ele)) :: elenodes
+      real, dimension(face_loc(field, face)) :: facenodes
+
+      integer :: i, j
+
+      face_rhs = 0.0
+      vector = 0.0
+
+      ! shape and detwei are the same for both faces, normal+ = - normal-
+      shape => face_shape(field, face)
+      call transform_facet_to_physical(X, face, detwei_f=detwei, normal=normal)
+
+      if (face==face_2) then  
+        ! boundary faces - need to apply weak dirichlet bc's
+        ! = - int_ v_h \cdot (u - u^b) n 
+        ! first check for weak-dirichlet bc
+        ! if no bc_info is applied then assume not weak bc
+        if (present(bc_type)) then
+          if (bc_type(face) == 1) then  
+            in_q = face_val_at_quad(field, face)
+            in_bc_q = ele_val_at_quad(bc_value, face)
+            
+            do i=1, mesh_dim(field)
+              vector(i,:) = -1.0*(in_q(:) - in_bc_q(:))*normal(i,:)
+            end do
+            face_rhs = shape_vector_rhs(shape, vector, detwei) 
+          end if
+        end if
+      else    
+        ! internal face
+        ! = int_ {v_h} \cdot J(x)  
+        in_q = face_val_at_quad(field, face)
+        in_q_2 = face_val_at_quad(field, face_2)
+
+        do i=1, mesh_dim(field)
+          vector(i,:) = -0.5*(in_q(:) - in_q_2(:))*normal(i,:)
+        end do
+        face_rhs = shape_vector_rhs(shape, vector, detwei) 
+      end if
+
+      elenodes = ele_nodes(field, ele)
```

Comment:
This whole block is unnecessarily complicated and inefficient (it's a face_loc*ele_loc loop within a loop over facets for each element). This can simply be replaced by:

lnodes => face_local_nodes(field, face)
loc_grad(:,lnodes) = loc_grad(:,lnodes) + face_rhs